### PR TITLE
dramatically upgrade function of --dateafter option with regards to c…

### PR DESCRIPTION
**Pull request by BenSapiens**
…ertain (bulk video) IEs known to return date-sorted playlists

## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

The code contained within this pull request, as stated in the commit message, dramatically upgrades the function of --dateafter with regards to playlists known to be date-ordered. For instance, we know that the values extracted by the YoutubePlaylistIE info extractor will be date-ordered if they originate with the YoutubeUserIE or YoutubeChannelIE info extractors, as both the YoutubeUserIE and YoutubeChannelIE info extractors, were they still used—not patched over by the solution to issue #5778—would return results in chronological order from most recent to least. By automatically not returning any further results after encountering a video date out of the date range object produced from date-restricting arguments passed by the user, we save youtube-dl from unnecessarily downloading the webpage of literally every single video beyond the dates that the user actually wants, which can number in the dozens or the hundreds.

TL;DR — I wrote this code because I wanted my "automatically get such-and-such users' new videos" script to download a total of 10 webpages instead of 1,000, and take 3 minutes to run rather than an hour.

### CODE REVIEW

File: youtube_dl/YoutubeDL.py [hunk 1]
```diff
@@ -758,6 +758,15 @@ def extract_info(self, url, download=True, ie_key=None, extra_info={},
                                     'and will probably not work.')
 
             try:
+                if self.original_ie:
+                    pass # self.original_ie already exists
+            except AttributeError:
+                try:
+                    self.original_ie = re.search("\.([^\.]+)'>", str(type(ie))).group(1)
```
Comment (dstftw) [inline youtube_dl/YoutubeDL.py]: This is a bad idea since it breaks reusing of `YoutubeDL` instances. Even worse it stores string instead of just reference.

File: youtube_dl/YoutubeDL.py [hunk 2]
```diff
@@ -758,6 +758,15 @@ def extract_info(self, url, download=True, ie_key=None, extra_info={},
...
+                except:
```
Comment (dstftw) [inline youtube_dl/YoutubeDL.py]: Never ever use bare except.

File: youtube_dl/YoutubeDL.py [hunk 3]
```diff
@@ -922,6 +931,8 @@ def iter_playlistitems(format):
 
             x_forwarded_for = ie_result.get('__x_forwarded_for_ip')
 
+            reliably_date_ordered_IEs = ('YoutubeChannelIE, YoutubeUserIE')
```
Comment (dstftw) [inline youtube_dl/YoutubeDL.py]: No, no hardcodes here. You are implementing a generic mechanism thus there should be a way to control it on per extractor basis. Or it would be easier to just introduce an option describing playlist order instead of this hack and require user to specify it if he wants such optimization.

File: youtube_dl/YoutubeDL.py [hunk 4]
```diff
@@ -948,7 +959,21 @@ def iter_playlistitems(format):
                 entry_result = self.process_ie_result(entry,
                                                       download=download,
                                                       extra_info=extra)
-                playlist_results.append(entry_result)
+
+                try:
+                    entry_result_date = entry_result['upload_date']
```
Comment (dstftw) [inline youtube_dl/YoutubeDL.py]: No `upload_date` is guaranteed to be present.

File: youtube_dl/YoutubeDL.py [hunk 5]
```diff
@@ -948,7 +959,21 @@ def iter_playlistitems(format):
...
+                    entry_result_date_year = int(entry_result_date[0:4])
+                    entry_result_date_month = int(entry_result_date[4:6])
+                    entry_result_date_day = int(entry_result_date[6:8])
+                    entry_result_date = datetime.date(year=entry_result_date_year, month=entry_result_date_month, day=entry_result_date_day)
+                    # if the entries originate with an info extractor known to return date-sorted results, simply break after we meet the first date out of range
+                    if self.original_ie in reliably_date_ordered_IEs and entry_result_date not in self.params['daterange']:
```
Comment (dstftw) [inline youtube_dl/YoutubeDL.py]: This won't work at all if starting playlist entry is not in `daterange`.

**Comment by BenSapiens** [comment]
Sorry for the flurry of minor changes. Hopefully the end result is to your satisfaction.

File: youtube_dl/options.py [hunk 6]
```diff
@@ -298,6 +298,10 @@ def _scrub_eq(o):
         metavar='DATE', dest='dateafter', default=None,
         help='Download only videos uploaded on or after this date (i.e. inclusive)')
     selection.add_option(
+        '--date-ordered-playlist',
```
Comment (dstftw) [inline youtube_dl/options.py]: Option should be called something like `--date-playlist-order` and should accept `asc`, `desc` or `none`. Code should process playlist according to it.

File: youtube_dl/YoutubeDL.py [hunk 7]
```diff
@@ -948,7 +950,14 @@ def iter_playlistitems(format):
                 entry_result = self.process_ie_result(entry,
                                                       download=download,
                                                       extra_info=extra)
+
+                entry_result_uploaddate = entry_result.get('upload_date')
+                if entry_result_uploaddate:
+                    if self.params.get('date_ordered_playlist') and entry_result_uploaddate not in self.params.get('daterange'):
```
Comment (dstftw) [inline youtube_dl/YoutubeDL.py]: I've already pointed out this does not work.

**Comment by BenSapiens** [comment]
Do you want something else done?

File: youtube_dl/options.py [hunk 8]
```diff
@@ -298,6 +298,15 @@ def _scrub_eq(o):
         metavar='DATE', dest='dateafter', default=None,
         help='Download only videos uploaded on or after this date (i.e. inclusive)')
     selection.add_option(
+        '--date-playlist-order',
+        metavar='ORDER', dest='date_playlist_order', default='none',
+        type='choice', choices=['asc', 'desc', 'none'],
+        help='Specify whether a playlist is known to be listed in chronological order. "Asc", "desc" or "none". Descending order is most recent to least, and is useful for batch-downloading the new videos from a Youtube channel. Default is "none", and is like not using this option at all. Causes youtube-dl to automatically stop after encountering the first video date outside of the date range configured by --date, --datebefore, and/or --dateafter.')
+    selection.add_option(
+        '--start-from-earliest',
```
Comment (dstftw) [inline youtube_dl/options.py]: There is already `--playlist-reverse`.

File: youtube_dl/options.py [hunk 9]
```diff
@@ -298,6 +298,15 @@ def _scrub_eq(o):
         metavar='DATE', dest='dateafter', default=None,
         help='Download only videos uploaded on or after this date (i.e. inclusive)')
     selection.add_option(
+        '--date-playlist-order',
+        metavar='ORDER', dest='date_playlist_order', default='none',
+        type='choice', choices=['asc', 'desc', 'none'],
+        help='Specify whether a playlist is known to be listed in chronological order. "Asc", "desc" or "none". Descending order is most recent to least, and is useful for batch-downloading the new videos from a Youtube channel. Default is "none", and is like not using this option at all. Causes youtube-dl to automatically stop after encountering the first video date outside of the date range configured by --date, --datebefore, and/or --dateafter.')
```
Comment (dstftw) [inline youtube_dl/options.py]: Should be cut down to coupe of sentences. Possible arguments' values should be written all lowercase without quotes. See `--fixup` for an example.

**Comment by BenSapiens** [comment]
Done and done, and also explicitly defined playlist_random and playlist_reverse, since every other option does it.

File: youtube_dl/YoutubeDL.py [hunk 10]
```diff
@@ -948,7 +953,14 @@ def iter_playlistitems(format):
                 entry_result = self.process_ie_result(entry,
                                                       download=download,
                                                       extra_info=extra)
+
+                entry_result_uploaddate = entry_result.get('upload_date')
```
Comment (dstftw) [inline youtube_dl/YoutubeDL.py]: Breaks on `-F`.

File: youtube_dl/YoutubeDL.py [hunk 11]
```diff
@@ -948,7 +953,14 @@ def iter_playlistitems(format):
...
+                if entry_result_uploaddate:
+                    if self.params.get('date_playlist_order') in ('desc', 'asc') and entry_result_uploaddate not in self.params.get('daterange'):
```
Comment (dstftw) [inline youtube_dl/YoutubeDL.py]: Did you even bother to run your code? `self.params.get('date_playlist_order')` will always be `None` thus `break` will never trigger since [you did not forward it to `params`](https://github.com/rg3/youtube-dl/blob/master/youtube_dl/__init__.py#L310-L429). Even if you did, it **won't work** (repeating this 3rd time already) if `daterange` is an **inner interval** inside `daterange`. In such case if the first/last video (depending on the order) does not belong to `daterange` extraction will stop on it and won't process further items that may belong to `daterange`.

File: youtube_dl/options.py [hunk 12]
```diff
@@ -482,11 +489,11 @@ def _scrub_eq(o):
         help=optparse.SUPPRESS_HELP)
     downloader.add_option(
         '--playlist-reverse',
-        action='store_true',
+        dest='playlist_reverse', action='store_true',
```
Comment (BenSapiens) [inline youtube_dl/options.py]: Okay.

File: youtube_dl/YoutubeDL.py (hunk 10 above)
Comment (BenSapiens) [inline youtube_dl/YoutubeDL.py]: Nice. Fixed.

File: youtube_dl/YoutubeDL.py (hunk 11 above)
Comment (BenSapiens) [inline youtube_dl/YoutubeDL.py]: > Did you even bother to run your code?

I've been running youtube-dl as described here: https://github.com/rg3/youtube-dl#embedding-youtube-dl, and completely forgot that I was manually specifying `ydl_opts`.

> In such case if the first/last video (depending on the order) does not belong to daterange extraction will stop on it and won't process further items that may belong to daterange.

That's literally the entire point.

Comment (dstftw) [inline youtube_dl/YoutubeDL.py]: No, it's not the point. It **must not** skip items than belong to the range specified.

**Comment by BenSapiens** [comment]
Alright...

**Comment by BenSapiens** [comment]
scratch that

**Comment by BenSapiens** [comment]
New commit is what preceding commit should have been.

Also, I wonder if, when encountering a "private video"  in a playlist, might it not be better to continue downloading the rest of the playlist?

Finally, I'm done. It's there; it works; hopefully all those people who asked you for this feature over the years will benefit.

In parting I would leave you with these words: "Any customer can have a car painted any color that he wants so long as it is black." — Henry Ford

**Comment by BenSapiens** [comment]
Alright. I'm back. What do you want to get this merged? I want the many other users of youtube-dl to benefit.
